### PR TITLE
Improve performance by searching substrings instead of regex.

### PR DESCRIPTION
**Pull request by bayleedev**
I'm using this plugin to search my bookmarks, on this laptop there are about 62 and I was noticing a large lag. I eventually narrowed it down to a single bookmarklet that took close to 2 seconds to query with a medium search term.

I decided to try out your `indexOf` idea by creating a `search(haystack).for(needle)` implementation. It's recursive, and I may have blinders on, but I think it's easy enough to read.

I made a benchmark test, but I'm not sure committing this would be ideal. But please reproduce locally to verify:

|  | all | item |
| --- | :-: | --: |
| before | 127,919ms | 1,279ms |
| after | 2ms | lawl |

``` js
var fuzzy = require('../')
var test = require('tape')

test('Long running match', function (t) {
  var totalDuration = 0
  for (let i = 0;i<100;i++) {
    var subset = [
      'Set Versionjavascript:(function()%7Bvar product%3Dwindow.prompt(%27What product%3F%27,%27filterable_errors%27)%3Bvar key %3D %27use_%27 %2B product %2B %27_version%27%3Bvar version%3Dwindow.prompt(%27What version%3F%27,%27hot%27)%3Bvar queries%3Dwindow.location.search.split(/%5B%5C%3F%7C%26%5D%2B/)%3Bqueries%3Dqueries.filter((query)%3D>%7Breturn query%7D)%3Bvar alreadyExists%3Dfalse%3Bqueries.forEach((query,index)%3D>%7Bif(query.includes(key))%7BalreadyExists%3Dtrue%3BqueryParts%3Dquery.split(%27%3D%27)%3BqueryParts%5B1%5D%3Dversion%3Bqueries%5Bindex%5D%3DqueryParts.join(%27%3D%27)%3B%7D%7D)%3Bif(!alreadyExists)%7Bqueries.push(key%2B%27%3D%27%2Bversion)%3B%7Dif(queries.length%3D%3D%3D1)%7Bwindow.location.search%3Dqueries%5B0%5D%3B%7Delse%7Bwindow.location.search%3Dqueries.join(%27%26%27)%3B%7D%7D)()%3B'
    ]
    t.plan(1)
    var start = new Date()
    fuzzy('browser-settings', subset)
    var end = new Date()
    totalDuration += end - start
  }
  t.ok(totalDuration < 10000, 'wanted ' + totalDuration + ' to be less than 10 seconds')
})
```

### CODE REVIEW

File: perf/index.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    var subset = [
+      'Set Versionjavascript:(function()%7Bvar product%3Dwindow.prompt(%27What product%3F%27,%27filterable_errors%27)%3Bvar key %3D %27use_%27 %2B product %2B %27_version%27%3Bvar version%3Dwindow.prompt(%27What version%3F%27,%27hot%27)%3Bvar queries%3Dwindow.location.search.split(/%5B%5C%3F%7C%26%5D%2B/)%3Bqueries%3Dqueries.filter((query)%3D>%7Breturn query%7D)%3Bvar alreadyExists%3Dfalse%3Bqueries.forEach((query,index)%3D>%7Bif(query.includes(key))%7BalreadyExists%3Dtrue%3BqueryParts%3Dquery.split(%27%3D%27)%3BqueryParts%5B1%5D%3Dversion%3Bqueries%5Bindex%5D%3DqueryParts.join(%27%3D%27)%3B%7D%7D)%3Bif(!alreadyExists)%7Bqueries.push(key%2B%27%3D%27%2Bversion)%3B%7Dif(queries.length%3D%3D%3D1)%7Bwindow.location.search%3Dqueries%5B0%5D%3B%7Delse%7Bwindow.location.search%3Dqueries.join(%27%26%27)%3B%7D%7D)()%3B'
+    ]
+    t.plan(1)
+    var start = new Date()
+    fuzzy('browser-settings', subset)
+    var end = new Date()
+    totalDuration += end - start
+  }
+  t.ok(totalDuration < 10, 'wanted ' + totalDuration + 'ms to be less than 10ms')
```
Comment (bayleedev) [inline perf/index.js:16]: I got 6ms locally for the 500 iterations. Travis did as well. Do you think `6<10` is too close? It's currently setup to fail the build if it is.

**Comment by bayleedev** [comment]
Also, given how fast this is, what are your thoughts on supporting a low n-gram type query?

I believe you are familiar, but when I type in `abcd` I'd do a search for `abcd`, `abc`, `bcd`, `ab`, `bc` and `cd`. Of course returning after the first match, preferring the larger match.

File: perf/index.js (hunk 1 above)
Comment (amjith) [inline perf/index.js:16]: We should `console.log()` the `totalDuration` so we can see how far off we were if we ever fail. 

10ms sounds good.

**Comment by amjith** [comment]
This is impressive. The code looks straight-forward and I like the search().for() semantic. 👍

**Comment by amjith** [comment]
I'm not opposed to the idea of having n-grams but I think it might start to screw with the ranking algorithm. 

I'd like it to be an optional feature to be turned on via options.

File: test/search.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  t.equal(result.end, 7, 'Expected to end at index 7')
+})
+
+test('test small search query', function (t) {
+  t.plan(3)
+  var result = search('hello world').for('w')
+
+  t.ok(result.found, 'Expected to find substring')
+  t.equal(result.start, 6, 'Expected to start at index 6')
+  t.equal(result.end, 7, 'Expected to end at index 7')
```
Comment (amjith) [inline test/search.js:35]: Shouldn't the `result.start` and `result.end` be the same value?

Comment (bayleedev) [inline test/search.js:35]: No, it's inclusive. You can do a `'hello world'.substring(6,7)` with the output. I intentionally increment by one [here](https://github.com/amjith/fuzzyfind/pull/8/files#diff-cdf48528144bffc7feaa2840475a728dR10)

File: perf/index.js (hunk 1 above)
Comment (bayleedev) [inline perf/index.js:16]: It outputs the message in the test output, if you now where to look for it.

**Pull request merged**